database/src/service/read.rs

```diff
@@ -228,57 +342,177 @@ fn map_request(
+        get_block_info(&block_height, &table_block_infos)?.block_hash,
+    ))
 }
 
 /// [`ReadRequest::BlockExtendedHeaderInRange`].
 #[inline]
 fn block_extended_header_in_range(
-    env: &Arc<ConcreteEnv>,
+    env: &ConcreteEnv,
     range: std::ops::Range<BlockHeight>,
 ) -> ResponseResult {
-    todo!()
+    let env_inner = env.env_inner();
+    let tx_ro = set_tx_ro!(env, env_inner);
+    let tables = set_tables!(env, env_inner, tx_ro);
+
+    let vec = range
+        .into_par_iter()
+        .map(|block_height| {
+            let (tx_ro, tables) = get_tx_ro_and_tables!(env_inner, tx_ro, tables);
+            get_block_extended_header_from_height(&block_height, tables)
+        })
+        .collect::<Result<Vec<ExtendedBlockHeader>, RuntimeError>>()?;
+
+    Ok(Response::BlockExtendedHeaderInRange(vec))
 }
 
 /// [`ReadRequest::ChainHeight`].
 #[inline]
-fn chain_height(env: &Arc<ConcreteEnv>) -> ResponseResult {
-    todo!()
+fn chain_height(env: &ConcreteEnv) -> ResponseResult {
+    let env_inner = env.env_inner();
+    let tx_ro = env_inner.tx_ro()?;
+    let table_block_heights = env_inner.open_db_ro::<BlockHeights>(&tx_ro)?;
+    let table_block_infos = env_inner.open_db_ro::<BlockInfos>(&tx_ro)?;
+
+    let top_height = top_block_height(&table_block_heights)?;
+    let block_hash = get_block_info(&top_height, &table_block_infos)?.block_hash;
+
+    Ok(Response::ChainHeight(top_height, block_hash))
 }
 
 /// [`ReadRequest::GeneratedCoins`].
 #[inline]
-fn generated_coins(env: &Arc<ConcreteEnv>) -> ResponseResult {
-    todo!()
+fn generated_coins(env: &ConcreteEnv) -> ResponseResult {
+    let env_inner = env.env_inner();
+    let tx_ro = env_inner.tx_ro()?;
+    let table_block_heights = env_inner.open_db_ro::<BlockHeights>(&tx_ro)?;
+    let table_block_infos = env_inner.open_db_ro::<BlockInfos>(&tx_ro)?;
+
+    let top_height = top_block_height(&table_block_heights)?;
+
+    Ok(Response::GeneratedCoins(cumulative_generated_coins(
+        &top_height,
+        &table_block_infos,
+    )?))
 }
 
 /// [`ReadRequest::Outputs`].
 #[inline]
-#[allow(clippy::needless_pass_by_value)] // TODO: remove me
-fn outputs(env: &Arc<ConcreteEnv>, map: HashMap<Amount, HashSet<AmountIndex>>) -> ResponseResult {
-    todo!()
+fn outputs(env: &ConcreteEnv, map: HashMap<Amount, HashSet<AmountIndex>>) -> ResponseResult {
+    let env_inner = env.env_inner();
+    let tx_ro = set_tx_ro!(env, env_inner);
+    let tables = set_tables!(env, env_inner, tx_ro);
+
+    // -> Result<(AmountIndex, OutputOnChain), RuntimeError>
+    let inner_map = |amount, amount_index| {
+        let (tx_ro, tables) = get_tx_ro_and_tables!(env_inner, tx_ro, tables);
+
+        let pre_rct_output_id = PreRctOutputId {
+            amount,
+            amount_index,
+        };
+
+        let output = get_output(&pre_rct_output_id, tables.outputs())?;
+        let output_on_chain = output_to_output_on_chain(&output, amount, tables.tx_unlock_time())?;
```

Comment:
This needs to handle getting RCT outputs as as well, currently it is only getting pre-rct outputs